Crash on malformed ldap syntax in keycloak_ldap_group_mapper.groups_ldap_filter

It looks like the plugin is not able to handle errors regarding syntax problems in groups_ldap_filter in keycloak_ldap_group_mapper:

`Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place
 
│ Error: Plugin did not respond
│ 
│   with module.keycloak_dev.keycloak_ldap_group_mapper.first_user_group,
│   on ../../modules/keycloak/main.tf line 396, in resource "keycloak_ldap_group_mapper" "first_user_group":
│  396: resource "keycloak_ldap_group_mapper" "first_user_group" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
 
Stack trace from the terraform-provider-keycloak_v3.1.1 plugin:
 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x6edc71]
 
goroutine 53 [running]:
github.com/hashicorp/go-version.(*Version).String(0x0, 0xc0003eb890, 0x2)
	github.com/hashicorp/go-version@v1.2.0/version.go:359 +0x51
github.com/hashicorp/go-version.(*Version).Compare(0x0, 0xc000331b30, 0xc0000ec500)
	github.com/hashicorp/go-version@v1.2.0/version.go:118 +0x2f
github.com/hashicorp/go-version.(*Version).GreaterThanOrEqual(...)
	github.com/hashicorp/go-version@v1.2.0/version.go:293
github.com/mrparkers/terraform-provider-keycloak/keycloak.(*KeycloakClient).VersionIsGreaterThanOrEqualTo(0xc00076c660, 0x115178f, 0x6, 0xf96380)
	github.com/mrparkers/terraform-provider-keycloak/keycloak/version.go:20 +0x65
github.com/mrparkers/terraform-provider-keycloak/provider.getLdapGroupMapperFromData(0xc00076c660, 0xc0000f7000, 0xc0001f16c0)
	github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:158 +0xabf
github.com/mrparkers/terraform-provider-keycloak/provider.resourceKeycloakLdapGroupMapperUpdate(0xc0000f7000, 0x114e020, 0xc00076c660, 0x0, 0xffffffffffffffff)
	github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:231 +0x4e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc0005e8580, 0x12fb060, 0xc000317600, 0xc0000f7000, 0x114e020, 0xc00076c660, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/helper/schema/resource.go:293 +0x88
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0005e8580, 0x12fb060, 0xc000317600, 0xc00027e310, 0xc0003b6de0, 0x114e020, 0xc00076c660, 0x0, 0x0, 0x0, ...)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/helper/schema/resource.go:395 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc00000cbc0, 0x12fb060, 0xc000317600, 0xc00027e150, 0xc00000cbc0, 0xc00000cbd0, 0x11a6070)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/internal/helper/plugin/grpc_provider.go:952 +0x8cf
github.com/hashicorp/terraform-plugin-sdk/v2/internal/tfplugin5._Provider_ApplyResourceChange_Handler.func1(0x12fb060, 0xc000317600, 0x1101400, 0xc00027e150, 0xc000317600, 0x103d3e0, 0x1, 0xc00009da00)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/internal/tfplugin5/tfplugin5.pb.go:3312 +0x89
github.com/hashicorp/terraform-plugin-sdk/v2/plugin.Serve.func3.1(0x12fb120, 0xc000398390, 0x1101400, 0xc00027e150, 0xc00009d9e0, 0xc00009da00, 0xc000244ba0, 0x5d30c6, 0x10c1a20, 0xc000398390)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/plugin/serve.go:76 +0x87
github.com/hashicorp/terraform-plugin-sdk/v2/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x11173c0, 0xc00000cbc0, 0x12fb120, 0xc000398390, 0xc00076dda0, 0xc00050aba0, 0x12fb120, 0xc000398390, 0xc000140400, 0x83c)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/internal/tfplugin5/tfplugin5.pb.go:3314 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002af880, 0x1305220, 0xc0000b3500, 0xc0000e5500, 0xc0001902d0, 0x19189e0, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.30.0/server.go:1171 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc0002af880, 0x1305220, 0xc0000b3500, 0xc0000e5500, 0x0)
	google.golang.org/grpc@v1.30.0/server.go:1494 +0xcc5
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0000402d0, 0xc0002af880, 0x1305220, 0xc0000b3500, 0xc0000e5500)
	google.golang.org/grpc@v1.30.0/server.go:834 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.30.0/server.go:832 +0x1fd
 
Error: The terraform-provider-keycloak_v3.1.1 plugin crashed!
 
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.`

I managed to reproduce it with v3.2.0 as well:

```
Stack trace from the terraform-provider-keycloak_v3.2.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x74cd31]

goroutine 73 [running]:
github.com/hashicorp/go-version.(*Version).String(0x0, 0xc0003ef1c0, 0x2)
        github.com/hashicorp/go-version@v1.3.0/version.go:371 +0x51
github.com/hashicorp/go-version.(*Version).Compare(0x0, 0xc0000927d0, 0xc0001ea500)
        github.com/hashicorp/go-version@v1.3.0/version.go:118 +0x2f
github.com/hashicorp/go-version.(*Version).GreaterThanOrEqual(...)
        github.com/hashicorp/go-version@v1.3.0/version.go:305
github.com/mrparkers/terraform-provider-keycloak/keycloak.(*KeycloakClient).VersionIsGreaterThanOrEqualTo(0xc000498000, 0x10309ae, 0x6, 0xec0120)
        github.com/mrparkers/terraform-provider-keycloak/keycloak/version.go:21 +0x65
github.com/mrparkers/terraform-provider-keycloak/provider.getLdapGroupMapperFromData(0xc000498000, 0xc000098c80, 0xc000159718)
        github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:158 +0xab8
github.com/mrparkers/terraform-provider-keycloak/provider.resourceKeycloakLdapGroupMapperUpdate(0xc000098c80, 0x102d2a0, 0xc000498000, 0x0, 0xffffffffffffffff)
        github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:231 +0x4e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc000263500, 0x11d31e8, 0xc00039c240, 0xc000098c80, 0x102d2a0, 0xc000498000, 0x0, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:352 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000263500, 0x11d31e8, 0xc00039c240, 0xc0002a0380, 0xc000786840, 0x102d2a0, 0xc000498000, 0x0, 0x0, 0x0, ...)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:464 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0001a6f60, 0x11d31e8, 0xc00039c240, 0xc000092780, 0xc00039c240, 0xfe5e40, 0xc0002d2600)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc0001cc9c0, 0x11d3290, 0xc00039c240, 0xc0002a00e0, 0xc0001cc9c0, 0xc0002d2630, 0xc00082bba0)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0xfe5e40, 0xc0001cc9c0, 0x11d3290, 0xc0002d2630, 0xc000836240, 0x0, 0x11d3290, 0xc0002d2630, 0xc000244d00, 0x814)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000b36c0, 0x11dba58, 0xc000102d80, 0xc0003a0000, 0xc0003644e0, 0x174eec0, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0000b36c0, 0x11dba58, 0xc000102d80, 0xc0003a0000, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000382270, 0xc0000b36c0, 0x11dba58, 0xc000102d80, 0xc0003a0000)
        google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd

Error: The terraform-provider-keycloak_v3.2.0 plugin crashed!
```

--- a/keycloak/keycloak_client_test.go
+++ b/keycloak/keycloak_client_test.go
@@ -61,7 +61,7 @@ func TestAccKeycloakApiClientRefresh(t *testing.T) {
 	}
 
 	// skip test if running 12.x or greater
-	if v, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(Version_12); !v {
+	if v, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(Version_12); v {
 		t.Skip()
 	}
 

--- a/keycloak/keycloak_client_test.go
+++ b/keycloak/keycloak_client_test.go
@@ -61,7 +61,7 @@ func TestAccKeycloakApiClientRefresh(t *testing.T) {
 	}
 
 	// skip test if running 12.x or greater
-	if keycloakClient.VersionIsGreaterThanOrEqualTo(Version_12) {
+	if v, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(Version_12); !v {
 		t.Skip()
 	}
 

--- a/keycloak/version.go
+++ b/keycloak/version.go
@@ -15,8 +15,18 @@ const (
 	Version_13 Version = "13.0.0"
 )
 
-func (keycloakClient *KeycloakClient) VersionIsGreaterThanOrEqualTo(versionString Version) bool {
-	v, _ := version.NewVersion(string(versionString))
+func (keycloakClient *KeycloakClient) VersionIsGreaterThanOrEqualTo(versionString Version) (bool, error) {
+	if keycloakClient.version == nil {
+		err := keycloakClient.login()
+		if err != nil {
+			return false, err
+		}
+	}
 
-	return keycloakClient.version.GreaterThanOrEqual(v)
+	v, err := version.NewVersion(string(versionString))
+	if err != nil {
+		return false, nil
+	}
+
+	return keycloakClient.version.GreaterThanOrEqual(v), nil
 }

--- a/provider/resource_keycloak_ldap_group_mapper.go
+++ b/provider/resource_keycloak_ldap_group_mapper.go
@@ -120,7 +120,7 @@ func resourceKeycloakLdapGroupMapper() *schema.Resource {
 	}
 }
 
-func getLdapGroupMapperFromData(keycloakClient *keycloak.KeycloakClient, data *schema.ResourceData) *keycloak.LdapGroupMapper {
+func getLdapGroupMapperFromData(keycloakClient *keycloak.KeycloakClient, data *schema.ResourceData) (*keycloak.LdapGroupMapper, error) {
 	var groupObjectClasses []string
 
 	for _, groupObjectClass := range data.Get("group_object_classes").([]interface{}) {
@@ -154,15 +154,19 @@ func getLdapGroupMapperFromData(keycloakClient *keycloak.KeycloakClient, data *s
 		MappedGroupAttributes:           mappedGroupAttributes,
 		DropNonExistingGroupsDuringSync: data.Get("drop_non_existing_groups_during_sync").(bool),
 	}
+	versionOk, err := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_11)
+	if err != nil {
+		return nil, err
+	}
 
-	if groupsPath, ok := data.GetOk("groups_path"); ok && keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_11) {
+	if groupsPath, ok := data.GetOk("groups_path"); ok && versionOk {
 		mapper.GroupsPath = groupsPath.(string)
 	}
 
-	return mapper
+	return mapper, nil
 }
 
-func setLdapGroupMapperData(keycloakClient *keycloak.KeycloakClient, data *schema.ResourceData, ldapGroupMapper *keycloak.LdapGroupMapper) {
+func setLdapGroupMapperData(keycloakClient *keycloak.KeycloakClient, data *schema.ResourceData, ldapGroupMapper *keycloak.LdapGroupMapper) error {
 	data.SetId(ldapGroupMapper.Id)
 
 	data.Set("name", ldapGroupMapper.Name)
@@ -184,17 +188,27 @@ func setLdapGroupMapperData(keycloakClient *keycloak.KeycloakClient, data *schem
 	data.Set("mapped_group_attributes", ldapGroupMapper.MappedGroupAttributes)
 	data.Set("drop_non_existing_groups_during_sync", ldapGroupMapper.DropNonExistingGroupsDuringSync)
 
-	if ldapGroupMapper.GroupsPath != "" && keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_11) {
+	versionOk, err := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_11)
+	if err != nil {
+		return err
+	}
+
+	if ldapGroupMapper.GroupsPath != "" && versionOk {
 		data.Set("groups_path", ldapGroupMapper.GroupsPath)
 	}
+
+	return nil
 }
 
 func resourceKeycloakLdapGroupMapperCreate(data *schema.ResourceData, meta interface{}) error {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
 
-	ldapGroupMapper := getLdapGroupMapperFromData(keycloakClient, data)
+	ldapGroupMapper, err := getLdapGroupMapperFromData(keycloakClient, data)
+	if err != nil {
+		return err
+	}
 
-	err := keycloakClient.ValidateLdapGroupMapper(ldapGroupMapper)
+	err = keycloakClient.ValidateLdapGroupMapper(ldapGroupMapper)
 	if err != nil {
 		return err
 	}
@@ -204,7 +218,10 @@ func resourceKeycloakLdapGroupMapperCreate(data *schema.ResourceData, meta inter
 		return err
 	}
 
-	setLdapGroupMapperData(keycloakClient, data, ldapGroupMapper)
+	err = setLdapGroupMapperData(keycloakClient, data, ldapGroupMapper)
+	if err != nil {
+		return err
+	}
 
 	return resourceKeycloakLdapGroupMapperRead(data, meta)
 }
@@ -220,7 +237,10 @@ func resourceKeycloakLdapGroupMapperRead(data *schema.ResourceData, meta interfa
 		return handleNotFoundError(err, data)
 	}
 
-	setLdapGroupMapperData(keycloakClient, data, ldapGroupMapper)
+	err = setLdapGroupMapperData(keycloakClient, data, ldapGroupMapper)
+	if err != nil {
+		return err
+	}
 
 	return nil
 }
@@ -228,9 +248,12 @@ func resourceKeycloakLdapGroupMapperRead(data *schema.ResourceData, meta interfa
 func resourceKeycloakLdapGroupMapperUpdate(data *schema.ResourceData, meta interface{}) error {
 	keycloakClient := meta.(*keycloak.KeycloakClient)
 
-	ldapGroupMapper := getLdapGroupMapperFromData(keycloakClient, data)
+	ldapGroupMapper, err := getLdapGroupMapperFromData(keycloakClient, data)
+	if err != nil {
+		return err
+	}
 
-	err := keycloakClient.ValidateLdapGroupMapper(ldapGroupMapper)
+	err = keycloakClient.ValidateLdapGroupMapper(ldapGroupMapper)
 	if err != nil {
 		return err
 	}
@@ -240,7 +263,10 @@ func resourceKeycloakLdapGroupMapperUpdate(data *schema.ResourceData, meta inter
 		return err
 	}
 
-	setLdapGroupMapperData(keycloakClient, data, ldapGroupMapper)
+	err = setLdapGroupMapperData(keycloakClient, data, ldapGroupMapper)
+	if err != nil {
+		return err
+	}
 
 	return nil
 }

--- a/provider/resource_keycloak_ldap_group_mapper_test.go
+++ b/provider/resource_keycloak_ldap_group_mapper_test.go
@@ -262,7 +262,7 @@ func TestAccKeycloakLdapGroupMapper_updateLdapUserFederationInPlace(t *testing.T
 func TestAccKeycloakLdapGroupMapper_groupsPath(t *testing.T) {
 	t.Parallel()
 
-	if !keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_11) {
+	if ok, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_11); !ok {
 		t.Skip()
 	}
 

--- a/provider/resource_keycloak_openid_client_optional_scopes_test.go
+++ b/provider/resource_keycloak_openid_client_optional_scopes_test.go
@@ -13,7 +13,7 @@ import (
 
 // All openid clients in Keycloak will automatically have these scopes listed as "optional client scopes".
 func getPreAssignedOptionalClientScopes() []string {
-	if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6) {
+	if ok, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6); ok {
 		return []string{"address", "phone", "offline_access", "microprofile-jwt"}
 	} else {
 		return []string{"address", "phone", "offline_access"}
@@ -397,7 +397,7 @@ func testAccCheckKeycloakOpenidClientOptionalScopeIsNotAttached(resourceName, cl
 }
 
 func testKeycloakOpenidClientOptionalScopes_basic(client, clientScope string) string {
-	if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6) {
+	if ok, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6); ok {
 		return fmt.Sprintf(`
 data "keycloak_realm" "realm" {
 	realm = "%s"
@@ -512,7 +512,7 @@ resource "keycloak_openid_client_optional_scopes" "optional_scopes" {
 }
 
 func testKeycloakOpenidClientOptionalScopes_validationNoClient(client, clientScope string) string {
-	if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6) {
+	if ok, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6); ok {
 		return fmt.Sprintf(`
 data "keycloak_realm" "realm" {
 	realm = "%s"
@@ -565,7 +565,7 @@ resource "keycloak_openid_client_optional_scopes" "optional_scopes" {
 }
 
 func testKeycloakOpenidClientOptionalScopes_validationBearerOnlyClient(client, clientScope string) string {
-	if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6) {
+	if ok, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6); ok {
 		return fmt.Sprintf(`
 data "keycloak_realm" "realm" {
 	realm = "%s"
@@ -673,7 +673,7 @@ resource "keycloak_openid_client_optional_scopes" "optional_scopes" {
 }
 
 func testKeycloakOpenidClientOptionalScopes_duplicateScopeAssignment(client, clientScope string) string {
-	if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6) {
+	if ok, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_6); ok {
 		return fmt.Sprintf(`
 %s
 

--- a/provider/resource_keycloak_realm_events_test.go
+++ b/provider/resource_keycloak_realm_events_test.go
@@ -165,15 +165,15 @@ func TestAccKeycloakRealmEvents_unsetEnabledEventTypes(t *testing.T) {
 						}
 
 						//keycloak versions < 7.0.0 have 63 events, versions >=7.0.0 have 67 events, versions >=12.0.0 have 69 events
-						if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_13) {
+						if ok, _ := keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_13); ok {
 							if len(realmEventsConfig.EnabledEventTypes) != 77 {
 								return fmt.Errorf("exptected to enabled_event_types to contain all(77) event types, but it contains %d", len(realmEventsConfig.EnabledEventTypes))
 							}
-						} else if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_12) {
+						} else if ok, _ = keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_12); ok {
 							if len(realmEventsConfig.EnabledEventTypes) != 69 {
 								return fmt.Errorf("exptected to enabled_event_types to contain all(69) event types, but it contains %d", len(realmEventsConfig.EnabledEventTypes))
 							}
-						} else if keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_7) {
+						} else if ok, _ = keycloakClient.VersionIsGreaterThanOrEqualTo(keycloak.Version_7); ok {
 							if len(realmEventsConfig.EnabledEventTypes) != 67 {
 								return fmt.Errorf("exptected to enabled_event_types to contain all(67) event types, but it contains %d", len(realmEventsConfig.EnabledEventTypes))
 							}

